x-pack/plugin/logstash/src/test/java/org/elasticsearch/xpack/logstash/action/TransportGetPipelineActionTests.java

```diff
@@ -103,6 +115,55 @@ public void onFailure(Exception e) {
         }
     }
 
+    /**
+     * Test that the explicit and wildcard IDs are requested.
+     */
+    public void testGetPipelinesByExplicitAndWildcardIds() {
+        InternalSearchResponse internalSearchResponse = new InternalSearchResponse(prepareSearchHits(), null, null, null, false, null, 1);
+        SearchResponse searchResponse = new SearchResponse(
+            internalSearchResponse,
+            null,
+            1,
+            1,
+            0,
+            100,
+            ShardSearchFailure.EMPTY_ARRAY,
+            SearchResponse.Clusters.EMPTY,
+            null
+        );
+
+        SearchResponse mockResponse = mock(SearchResponse.class);
+        when(mockResponse.getHits()).thenReturn(prepareSearchHits());
+
+        GetPipelineRequest request = new GetPipelineRequest(List.of("1", "2", "3*"));
+        AtomicReference<Exception> failure = new AtomicReference<>();
+
+        // Set up an ActionListener for the actual test conditions
+        ActionListener<GetPipelineResponse> testActionListener = new ActionListener<>() {
+            @Override
+            public void onResponse(GetPipelineResponse getPipelineResponse) {
+                assertThat(getPipelineResponse, is(notNullValue()));
+                assertThat(getPipelineResponse.pipelines().size(), equalTo(3));
+                assertTrue(getPipelineResponse.pipelines().containsKey("1"));
+                assertTrue(getPipelineResponse.pipelines().containsKey("2"));
+                assertTrue(getPipelineResponse.pipelines().containsKey("3*"));
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                failure.set(e);
+            }
+        };
+
+        new TransportGetPipelineAction(mock(TransportService.class), mock(ActionFilters.class), getMockClient(searchResponse)).doExecute(
+            null,
+            request,
+            testActionListener
+        );
```

Comment:
```suggestion
        try (Client client = getMockClient(searchResponse)) {
            new TransportGetPipelineAction(mock(TransportService.class), mock(ActionFilters.class), client).doExecute(
                null,
                request,
                testActionListener
            );
        }
```

Reply:
Can you plz explain what is happening behind the scene?

Reply:
The thread leak detection is taking a snapshot of all threads currently created before the test and when the test completes. If there are any threads that are left lying around, it throws an assertion error. Hence the failures. 

The `getMockClient()` method is returning a new instance of `NoOpClient`. Inside of its constructor, it is creating a new instance of `TestThreadPool` which is a subclass of `ThreadPool`. The `ThreadPool` instance creates a number of executors, schedulers, and threads (like the time caching thread). All of those threads need to be shutdown before the end of the test case. Putting client in the try-with-resources block will ensure that the client is closed after the block, which will close the thread pool underneath.